components/lib/nearby/src/main/java/mozilla/components/lib/nearby/NearbyConnection.kt

```diff
@@ -0,0 +1,346 @@
+     */
+    fun startDiscovering() {
+        connectionsClient.startDiscovery(
+            PACKAGE_NAME, endpointDiscoveryCallback,
+            DiscoveryOptions.Builder().setStrategy(STRATEGY).build())
+            .addOnSuccessListener {
+                updateState(ConnectionState.Discovering)
+            }.addOnFailureListener {
+                reportError("failed to start discovering: $it")
+            }
+    }
+
+    // Discovery
+    private val endpointDiscoveryCallback = object : EndpointDiscoveryCallback() {
+        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
+            connectionsClient.requestConnection(name, endpointId, connectionLifecycleCallback)
+            updateState(ConnectionState.Connecting(endpointId, info.endpointName))
+        }
+
+        override fun onEndpointLost(endpointId: String) {
+            updateState(ConnectionState.Discovering)
+        }
+    }
+
+    // Used within startAdvertising() and startDiscovering() (via endpointDiscoveryCallback)
+    private val connectionLifecycleCallback = object : ConnectionLifecycleCallback() {
+        override fun onConnectionInitiated(endpointId: String, connectionInfo: ConnectionInfo) {
+            // This is the last time we have access to the endpoint name, so cache it.
+            endpointIdsToNames.put(endpointId, connectionInfo.endpointName)
+            if (authenticate) {
+                updateState(
+                    ConnectionState.Authenticating(
+                        this@NearbyConnection,
+                        endpointId,
+                        connectionInfo.endpointName,
+                        connectionInfo.authenticationToken
+                    )
+                )
+            } else {
+                connectionsClient.acceptConnection(endpointId, payloadCallback)
+                updateState(ConnectionState.Connecting(endpointId, connectionInfo.endpointName))
+            }
+        }
+
+        override fun onConnectionResult(endpointId: String, result: ConnectionResolution) {
+            if (result.status.isSuccess) {
+                connectionsClient.stopDiscovery()
+                connectionsClient.stopAdvertising()
+                updateState(ConnectionState.ReadyToSend(
+                    endpointId,
+                    endpointIdsToNames[endpointId]))
+            } else {
+                reportError("onConnectionResult: connection failed with status ${result.status}")
+            }
+        }
+
+        override fun onDisconnected(endpointId: String) {
+            updateState(ConnectionState.Isolated)
+        }
+    }
+
+    private val payloadCallback = object : PayloadCallback() {
+        override fun onPayloadReceived(endpointId: String, payload: Payload) {
+            listener?.receiveMessage(
+                endpointId,
+                endpointIdsToNames[endpointId],
+                String(payload.asBytes()!!, UTF_8))
```

Comment:
`!!` isn't a safe operation. Probably best to null-check the `asBytes` call.

Reply:
Yeah, I shouldn't have done that. Fixed.